cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -388,22 +399,78 @@ where
 	/// Checks the execution receipt from the executor peers.
 	fn on_execution_receipt(
 		&self,
-		_execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
+		execution_receipt: &ExecutionReceipt<<Block as BlockT>::Hash>,
 	) -> Result<Action, Self::Error> {
 		// TODO: validate the Proof-of-Election
 
-		// TODO: check if the received ER is same with the one produced locally.
-		let same_with_produced_locally = true;
-
-		if same_with_produced_locally {
-			Ok(Action::RebroadcastExecutionReceipt)
+		// TODO: more efficient execution receipt checking strategy?
+		let local_receipt = if let Some(local_receipt) =
+			crate::aux_schema::load_execution_receipt::<_, Block>(
+				&*self.client,
+				execution_receipt.secondary_hash,
+			)? {
+			local_receipt
 		} else {
+			// Wait for the local execution receipt until it's ready.
+			let (tx, rx) = crossbeam::channel::bounded::<
+				sp_blockchain::Result<ExecutionReceipt<Block::Hash>>,
+			>(1);
```

Comment:
Is this the case for a oneshot channel?
https://docs.rs/tokio/1.10.0/tokio/sync/oneshot/index.html

Reply:
Or is it here in order to block on it in sync context and not to block in async future?

Reply:
Yeah, it's the oneshot case. The problem here is how to properly get a value asynchronously under a sync context. The background is that when an executor received an execution receipt from the network, the local execution receipt might be not ready and you have to wait because you can't check the external execution receipt until you have generated it on your own locally.

Reply:
I think it is acceptable for now.

I traced call stack and it is called from manually written future in Substrate's source code. So while we could have refactored everything all the way up there to make it async, it is unlikely to be accepted and from one of the PRs Liu-Cheng sent upstream it looks like there is a chance this gossiping mechanism will go away eventually.

We'll likely switch to our own networking layer for these kinds of things once it is ready.

Just make sure this doesn't cause disasters as it does seem to block networking completely.